### ISSUE TITLE
Forbid direct instantiation via "new" keyword (or, enforce factory method)

### ISSUE THREAD

**Issue by TravisCarden** [slevomat/coding-standard#1567]
# Feature request

_Cross-posted from https://github.com/phpstan/phpstan/issues/9264:_

**tl;dr:** I would like to forbid direct instantiation of a certain class via the `new` keyword, e.g., `new Example()`, so instances of it can only be created via an injectable factory service. Perhaps such a rule exists somewhere--which would be fantastic--but I can't find one.

**Rationale:** I have a requirement for all user-facing strings to be translatable (i18n/l10n) by Drupal (https://localize.drupal.org/). The way that platform works is that it statically analyzes code for calls to a `t()` method and extracts the literal string arguments. I have a `TranslatableMessage` class to enforce type-hinting so no bare strings can slip through, and I have a factory class that instantiates it via the requisite `t()` method. Now I just need to make sure that no code unwittingly creates these classes directly--which would conform to type-hinted interfaces and thus pass automated tests but would prevent Drupal.org from extracting them. I've explored some other design approaches to solving the problem, but all of them have undesirable side effects, like preventing dependent projects from substituting their own translatable message implementations or making return type hints impossible. I'd really like to have a PHPCS rule that lets me specify forbidden classes in my `phpcs.xml` and just suppress it on their factories.

**Comment by TravisCarden** [comment]
Ah, it looks like [`SlevomatCodingStandard.PHP.ForbiddenClasses`](https://github.com/slevomat/coding-standard/blob/master/doc/php.md#slevomatcodingstandardphpforbiddenclasses-) is what I was looking for. Thanks!
